### PR TITLE
Log and expose cgroup OOM events to the associated Pod resource

### LINKED ISSUE #100483: Log and expose cgroup OOM events to the associated Pod resource
#### What happened:

Previously the cadvisor library had a regex parsing error, which resulted in it falling back to returning
a string character of `/` when parsing system oom messages on kernels >= 5.0. This was fixed in google/cadvisor#2813

~Since cadvisor was bumped in #99875, system OOM event messages are no longer being emitted because of the following line, where its checking that the `event.VictimContainerName == "/"`.
https://github.com/kubernetes/kubernetes/blob/e557f61784a90adf8dfe4a0bca875043e895cc8b/pkg/kubelet/oom/oom_watcher_linux.go#L75~

Since cadvisor was bumped in #99875, we can now retrieve the OOM'd pod id, creating a log entry and emit an event for the associated resource

#### What you expected to happen:

System OOM messages events to be emitted to the Node resource.

#### How to reproduce it (as minimally and precisely as possible):

Running a 1.21.x cluster, create a pod that OOM's.

```
apiVersion: v1
kind: Pod
metadata:
  name: memory-demo-2
  namespace: default
spec:
  containers:
  - name: memory-demo-2-ctr
    image: polinux/stress
    resources:
      requests:
        memory: "50Mi"
      limits:
        memory: "100Mi"
    command: ["stress"]
    args: ["--vm", "1", "--vm-bytes", "250M", "--vm-hang", "1"]
```

Observe that there is no SystemOOM message for the events of the node where that pod is running.

#### Environment:
- Kubernetes version (use `kubectl version`): v1.21.0-beta.1.382+1a983bb958ba66
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`): `Linux ip-172-31-48-224 5.4.0-1038-aws #40-Ubuntu SMP Fri Feb 5 23:50:40 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by kgtw**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

With cadvisor v0.39.0, the OOM parser correctly returns the container information for kernels >= 5.0. This PR has two aspects.

This allows us to expose the underlying cgroup OOM to the associated Pod resource by emitting an event. In addition, we should log the cgroup OOM in the kubelet logs.
- Adds cgroup OOM log message to Kubelet
- Adds the cgroup OOM event messages to the associated Pod resource.

#### Which issue(s) this PR fixes:

Fixes #100483

#### Special notes for your reviewer:

```
$ kubectl describe pod -n default memory-demo-2
...
Events:
  Type     Reason     Age                From               Message
  ----     ------     ----               ----               -------
  Normal   Scheduled  30s                default-scheduler  Successfully assigned default/memory-demo-2 to 127.0.0.1
  Normal   Pulled     28s                kubelet            Successfully pulled image "polinux/stress" in 1.145060952s
  Warning  CgroupOOM  28s                kubelet            Cgroup OOM encountered, victim process: stress, pid: 470521
```

```
$ kubectl describe node 127.0.0.1
...
Events:
  Type     Reason     Age                 From        Message
  ----     ------     ----                ----        -------
  Normal   Starting   5m28s               kube-proxy  Starting kube-proxy.
```
#### Does this PR introduce a user-facing change?

```release-note
Emit pod CgroupOOM events to the associated Pod resource
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by kgtw** [comment]
Following up on discussions on the issue https://github.com/kubernetes/kubernetes/issues/100483#issuecomment-806224383, I will be changing this PR from a bug/regression to a feature.

/remove-kind bug
/remove-kind regression
/kind feature

**Comment by ravisantoshgudimetla** [comment]
cc @rphillips @mrunalp @sjenning

File: pkg/kubelet/oom/oom_watcher_linux.go [hunk 1]
```diff
@@ -59,25 +65,38 @@ func NewWatcher(recorder record.EventRecorder) (Watcher, error) {
...
+				eventMsg = "Cgroup OOM encountered"
+			}
+			if event.ProcessName != "" && event.Pid != 0 {
+				eventMsg = fmt.Sprintf("%s, victim process: %s, pid: %d", eventMsg, event.ProcessName, event.Pid)
+			}
+
+			parsedContainer := containerRegexp.FindStringSubmatch(event.VictimContainerName)
+			if parsedContainer != nil && event.ContainerName != legacyRecordEventContainerName {
+				klog.V(1).InfoS("Got cgroup oom event", "event", event)
+				if pod, ok := getPodByUID(types.UID(parsedContainer[1])); ok {
```
Comment (rphillips) [inline pkg/kubelet/oom/oom_watcher_linux.go:93]: parsedContainer at index 1 is not guaranteed to be set with FindStringSubmatch. We need a check on line 91 to make sure it exists.

Comment (kgtw) [reply pkg/kubelet/oom/oom_watcher_linux.go:93]: Good catch, I've fixed this in the latest commit.

**Comment by bobbypage** [comment]
Thanks for the PR, I agree it would great to have events for cgroup OOMs.

I'm not a big fan of using the same approach of relying on scraping dmesg logs to get OOMs. As we've seen in the past, this can be fragile and the log format change across kernel versions.

For system OOMs that's the only option as far as I'm aware, but for cgroup OOMs there's a much more robust option: the kernel can report back when a cgroup OOMs via eventfd. See section 10 here: https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt

I believe the container runtime is already using this capability, for example containerd CRI will set the status to `OOMKilled`: 

 in case of OOM kill, see [here](https://github.com/containerd/containerd/blob/4f181312393502a07b298b73dcdd379832612ecd/pkg/cri/server/events.go#L334-L350). This is all detected via the existing eventfd mechanism described above, implemented [here](https://github.com/containerd/containerd/blob/4f18131/pkg/oom/v1/v1.go#L92)

Dockershim is also doing this and setting container status reason to `OOMKilled`: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockershim/docker_container.go#L415

So I think the more robust approach here is to rely on the runtime to detect OOM Kill as opposed to scraping the dmesg logs. When kubelet finds out a container status has been updated with `OOMKilled` reason from the CRI, report that as a k8s event.

**Comment by prometheus-tao** [comment]
kubelet log show  system oom event , The Node occur not ready ?

**Pull request closed without merging**